evm/rlp/headers.py

```diff
@@ -179,83 +179,135 @@ def create_execution_context(self, prev_hashes):
+                int_to_bytes32(self.proposer_bid),
+            ])
         )
-        return header_hash
+
+    def to_signed_collation(self):
+        raise NotImplementedError("TODO")
+
+
+class CollationHeader(rlp.Serializable):
+    """The header of a collation signed by the proposer."""
+
+    fields = [
+        ("shard_id", int32),
+        ("parent_hash", hash32),
+        ("chunks_root", hash32),
+        ("period", int32),
+        ("height", int32),
+        ("proposer_address", address),
+        ("proposer_bid", int32),
+        ("proposer_signature", Binary.fixed_length(96)),
+    ]
+
+    smc_field_sizes = [
+        ("shard_id", 32),
+        ("parent_hash", 32),
+        ("chunks_root", 32),
+        ("period", 32),
+        ("height", 32),
+        ("proposer_address", 32),
+        ("proposer_bid", 32),
+        ("proposer_signature", 96),
+    ]
+    smc_encoded_size = sum(size for _, size in smc_field_sizes)
+    assert [name for name, _ in smc_field_sizes] == [name for name, _ in fields]
+
+    def __repr__(self):
+        return "<CollationHeader #{0} {1} (shard #{2})>".format(
+            self.height,
+            encode_hex(self.hash)[2:10],
+            self.shard_id,
+        )
+
+    def to_bytes(self):
+        encoded = b''.join([
+            int_to_bytes32(self.shard_id),
+            self.parent_hash,
+            self.chunks_root,
+            int_to_bytes32(self.period),
+            int_to_bytes32(self.height),
+            pad32(self.proposer_address),
+            int_to_bytes32(self.proposer_bid),
+            self.proposer_signature,
+        ])
+        assert len(encoded) == self.smc_encoded_size
+        return encoded
+
+    @property
+    def hash(self):
+        return keccak(self.to_bytes())
 
     @classmethod
     @to_dict
     def _deserialize_header_bytes_to_dict(cls, header_bytes):
-        # assume all fields are padded to 32 bytes
-        obj_size = 32
-        if len(header_bytes) != obj_size * len(cls.fields):
+        if len(header_bytes) != cls.smc_encoded_size:
             raise ValidationError(
-                "Expected header bytes to be of length: {0}. Got length {1} instead.\n- {2}".format(
-                    obj_size * len(cls.fields),
+                "Expected encoded header to be of size: {0}. Got size {1} instead.\n- {2}".format(
+                    cls.smc_encoded_size,
                     len(header_bytes),
                     encode_hex(header_bytes),
                 )
             )
-        for idx, field in enumerate(cls.fields):
-            field_name, field_type = field
-            start_index = idx * obj_size
-            field_bytes = header_bytes[start_index:(start_index + obj_size)]
+
+        start_indices = accumulate(lambda i, field: i + field[1], cls.smc_field_sizes, 0)
+        start_index = next(start_indices)
+        for field_name, field_type in cls.fields:
+            end_index = next(start_indices)
```

Comment:
I think this loop can be changed as follows.

```
from cytoolz import sliding_window
field_bounds = sliding_window(2, start_indices)
for ((start_idx, end_idx), (field_name, field_type)) in zip(cls.fields, field_bounts):
    ...
```

Removes mutability as well as extracting the start/end index logic out of the loop body which I think improves readability some.